test/test_function_schema.py

```diff
@@ -10,8 +11,81 @@ def test_serialize_and_deserialize(self):
         # so far we have around 1700 registered schemas
         self.assertGreater(len(schemas), 1000)
         for schema in schemas:
-            parsed_schema = torch._C.parse_schema(str(schema))
+            parsed_schema = parse_schema(str(schema))
             self.assertEqual(parsed_schema, schema)
+            self.assertTrue(parsed_schema.is_backcompat_with(schema))
+
+    def test_backward_compatible_optional_arg(self):
+        old_schema = parse_schema('any(Tensor self, int dim, bool keepdim=False) -> Tensor')
+        new_schema = parse_schema('any(Tensor self, int? dim, bool keepdim=False) -> Tensor')
+        self.assertTrue(new_schema.is_backcompat_with(old_schema))
+        self.assertFalse(old_schema.is_backcompat_with(new_schema))
+
+    def test_backward_compatible_default_value(self):
+        old_schema = parse_schema('any(Tensor self, int dim, bool keepdim=False) -> Tensor')
+        new_schema = parse_schema('any(Tensor self, int dim=5, bool keepdim=False) -> Tensor')
+        self.assertTrue(new_schema.is_backcompat_with(old_schema))
+        self.assertFalse(old_schema.is_backcompat_with(new_schema))
+
+    def test_backward_compatible_add_args(self):
+        old_schema = parse_schema('any(Tensor self, int dim) -> Tensor')
+        new_schema = parse_schema('any(Tensor self, int dim, bool keepdim=False) -> Tensor')
+        self.assertTrue(new_schema.is_backcompat_with(old_schema))
+        self.assertFalse(old_schema.is_backcompat_with(new_schema))
+        new_schema = parse_schema('any(Tensor self, int dim, bool? keepdim) -> Tensor')
+        self.assertTrue(new_schema.is_backcompat_with(old_schema))
+        self.assertFalse(old_schema.is_backcompat_with(new_schema))
+
+    def test_backward_incompatible_name(self):
+        old_schema = parse_schema('any(Tensor self, int dim, bool keepdim=False) -> Tensor')
+        new_schema = parse_schema('any_(Tensor self, int dim, bool keepdim=False) -> Tensor')
+        self.assertFalse(new_schema.is_backcompat_with(old_schema))
+        self.assertFalse(old_schema.is_backcompat_with(new_schema))
+
+    def test_backward_incompatible_vararg(self):
+        old_schema = parse_schema('any(Tensor self, int dim, bool keepdim=False) -> Tensor')
+        new_schema = parse_schema('any(Tensor self, int dim, bool keepdim=False, ...) -> Tensor')
+        self.assertFalse(new_schema.is_backcompat_with(old_schema))
+        self.assertFalse(old_schema.is_backcompat_with(new_schema))
+
+    def test_backward_incompatible_varret(self):
+        old_schema = parse_schema('any(Tensor self, int dim, bool keepdim=False) -> Tensor')
+        new_schema = parse_schema('any(Tensor self, int dim, bool keepdim=False) -> (Tensor, ...)')
+        self.assertFalse(new_schema.is_backcompat_with(old_schema))
+        self.assertFalse(old_schema.is_backcompat_with(new_schema))
+
+    def test_backward_incompatible_returns(self):
```

Comment:
Technically it is valid for someone to change a return of Tensor? to Tensor and it is backward compatible.

Reply:
it's not BC is C++ and even in torch script - don't we generate some handling IR around if it's optional? That being said, we don't currently have optional outputs afaiu, so it's fine to be stricter

Reply:
IR is re-generated on load since we are loading from python, so any optional-specific IR won't get generated. This happens in other places as well. For instance, when we inline we are replacing a T? with a T all the time.

Reply:
yes, i added the handling for this case.